src/util/hashFetch/index.js

```diff
@@ -0,0 +1,247 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+const fs = require('fs');
+const util = require('util');
+const path = require('path');
+
+import unzip from 'unzipper';
+
+import { getHashFetchPath } from '../host';
+import ExpoRetry from './expoRetry';
+import Contracts from '@parity/shared/lib/contracts';
+import { sha3 } from '@parity/api/lib/util/sha3';
+import { bytesToHex } from '@parity/api/lib/util/format';
+import { ensureDir as fsEnsureDir, emptyDir as fsEmptyDir } from 'fs-extra';
+import { http, https } from 'follow-redirects';
+
+const fsExists = util.promisify(fs.stat);
+const fsStat = util.promisify(fs.stat);
+const fsRename = util.promisify(fs.rename);
+const fsReadFile = util.promisify(fs.readFile);
+const fsUnlink = util.promisify(fs.unlink);
+const fsReaddir = util.promisify(fs.readdir);
+const fsRmdir = util.promisify(fs.rmdir);
+
+const MAX_DOWNLOADED_FILE_SIZE_BYTES = 10485760; // 20MB
+
+function registerFailedAttemptAndThrow (hash, url, e) {
+  ExpoRetry.get().registerFailedAttempt(hash, url);
+  throw e;
+}
+
+function checkHashMatch (hash, path) {
+  return fsReadFile(path).then(content => {
+    if (sha3(content) !== `0x${hash}`) { throw new Error(`Hashes don't match: expected 0x${hash}, got ${sha3(content)}`); }
+  });
+}
+
+// List directory contents and tell if each item is a file or a directory
+function ls (folderPath) {
+  return fsReaddir(folderPath)
+    .then(filenames =>
+      Promise.all(filenames.map(filename => {
+        const filePath = path.join(folderPath, filename);
+
+        return fsStat(filePath).then(stat => ({ isDirectory: stat.isDirectory(), filePath, filename }));
+      }))
+    );
+}
+
+function rawUnzipTo (zipPath, extractPath) {
+  return new Promise((resolve, reject) => {
+    const unzipParser = unzip.Extract({ path: extractPath });
+
+    fs.createReadStream(zipPath).pipe(unzipParser);
+
+    unzipParser.on('error', reject);
+
+    unzipParser.on('close', resolve);
+  });
+}
+
+function unzipThroughTo (tempPath, extractPath, finalPath) {
+  return rawUnzipTo(tempPath, extractPath)
+    .then(() => fsUnlink(tempPath))
+    .then(() => ls(extractPath))
+    .then(files => {
+      // Check if the zip file contained a root folder
+      if (files.length === 1 && files[0].isDirectory) {
+        // Rename the root folder (contaning the dapp) to finalPath
+        const rootFolderPath = files[0].filePath;
+
+        return fsRename(rootFolderPath, finalPath)
+            .then(() => fsRmdir(extractPath));
+      } else {
+        // No root folder: extractPath contains the dapp
+        return fsRename(extractPath, finalPath);
+      }
+    });
+}
+
+function download (url, destinationPath) {
```

Comment:
What will happen if the url is `file:///etc/passwd`?

Reply:
`download` is only called by `hashDownload`, which is called with a url that is guaranteed to start with http:
https://github.com/parity-js/shell/blob/f76f89cdde20604c68d508788691631bd8a0267b/src/util/hashFetch/index.js#L161-L172

I'll add a condition in the download() function for future-proofing